File: kotlin-node/karakum/src/jsMain/resources/injections/EventMapInjection.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

const eventHandlerMethods = new Set([
    "on",
    "off",
    "once",
    "emit",
    "addListener",
    "removeListener",
    "prependListener",
    "prependOnceListener",
    "listeners",
])

const openEvents = {}

const overriddenEvents = {
    "fs.d.ts": {
        "ReadStream": [
            "close",
            "data",
            "end",
            "error",
            "pause",
            "readable",
            "resume",
        ],
        "WriteStream": [
            "close",
            "drain",
            "error",
            "finish",
            "pipe",
            "unpipe",
        ],
    },
}

const ignoredEvents = {}

function extractEventContainer(node) {
    if (
        node.text === "K"

        && node.parent
        && ts.isTypeReferenceNode(node.parent)

        && node.parent.parent
        && ts.isParameter(node.parent.parent)
        && ts.isIdentifier(node.parent.parent.name)
        && node.parent.parent.name.text === "event"

        && node.parent.parent.parent
        && (
            (
                ts.isMethodDeclaration(node.parent.parent.parent)
                && ts.isIdentifier(node.parent.parent.parent.name)
                && eventHandlerMethods.has(node.parent.parent.parent.name.text)

                && node.parent.parent.parent.parent
                && ts.isClassDeclaration(node.parent.parent.parent.parent)
            )
            || (
                ts.isMethodSignature(node.parent.parent.parent)
                && ts.isIdentifier(node.parent.parent.parent.name)
                && eventHandlerMethods.has(node.parent.parent.parent.name.text)

                && node.parent.parent.parent.parent
                && ts.isInterfaceDeclaration(node.parent.parent.parent.parent)
            )
        )
    ) {
        return node.parent.parent.parent.parent
    }

    return null
}

function extractEventPayloads(node, context) {
    if (
        node.text === "K"

        && node.parent
        && ts.isTypeReferenceNode(node.parent)

        && node.parent.parent
        && ts.isParameter(node.parent.parent)
        && ts.isIdentifier(node.parent.parent.name)
        && node.parent.parent.name.text === "event"

        && node.parent.parent.parent
        && (
            (
                ts.isMethodDeclaration(node.parent.parent.parent)
                && ts.isIdentifier(node.parent.parent.parent.name)
                && node.parent.parent.parent.name.text === "addListener"

                && node.parent.parent.parent.parent
                && ts.isClassDeclaration(node.parent.parent.parent.parent)
            )
            || (
                ts.isMethodSignature(node.parent.parent.parent)
                && ts.isIdentifier(node.parent.parent.parent.name)
                && node.parent.parent.parent.name.text === "addListener"

                && node.parent.parent.parent.parent
                && ts.isInterfaceDeclaration(node.parent.parent.parent.parent)
            )
        )
    ) {
        const method = node.parent.parent.parent
        const typeParameter = method.typeParameters?.find(it => it.name.text === "K")

        if (
            typeParameter
            && typeParameter.constraint
            && ts.isTypeOperatorNode(typeParameter.constraint)
            && ts.isTypeReferenceNode(typeParameter.constraint.type)
        ) {
            const typeScriptService = context.lookupService(karakum.typeScriptServiceKey.get())
            const typeChecker = typeScriptService?.program.getTypeChecker()

            const symbol = typeChecker?.getSymbolAtLocation(typeParameter.constraint.type.typeName)

            if (
                symbol
                && symbol.declarations
                && symbol.declarations[0]
                && ts.isTypeAliasDeclaration(symbol.declarations[0])
                && ts.isIntersectionTypeNode(symbol.declarations[0].type)
            ) {
                const eventMap = symbol.declarations[0].type.types.find(it => ts.isTypeLiteralNode(it))

                if (
                    eventMap
                    && ts.isTypeLiteralNode(eventMap)
                ) {
                    return eventMap.members.reduce((acc, member) => {
                        if (
                            ts.isPropertySignature(member)
                            && ts.isIdentifier(member.name)

                            && member.type
                            && ts.isFunctionTypeNode(member.type)
                        ) {
                            acc.set(member.name.text, member.type.parameters)
                        }

                        return acc
                    }, new Map())
                }
            }
        }

        const typeScriptService = context.lookupService(karakum.typeScriptServiceKey.get())

        console.error(`Suspicious listener: ${typeScriptService?.printNode(listener)}`)
    }

    return null
}

export default {
    setup(context) {
        this.events = new Map()
    },

    traverse(node, context) {
        if (ts.isIdentifier(node)) {
            const eventContainer = extractEventContainer(node)
            const eventPayloads = extractEventPayloads(node, context)

            if (!eventContainer) return null
            if (!eventPayloads) return null

            const name = eventContainer.name
            if (!name) return null

            const typeScriptService = context.lookupService(karakum.typeScriptServiceKey.get())
            const typeChecker = typeScriptService?.program.getTypeChecker()

            const symbol = typeChecker?.getSymbolAtLocation(name)
            if (!symbol) return

            this.events.set(symbol, eventPayloads)
        }
    },

    render(node) {
        if (
            (
                ts.isMethodDeclaration(node)
                || ts.isMethodSignature(node)
            )
            && ts.isIdentifier(node.name)
            && eventHandlerMethods.has(node.name.text)

            && node.parameters[0]
            && ts.isIdentifier(node.parameters[0].name)
            && node.parameters[0].name.text === "event"

            && node.parameters[0].type
            && ts.isTypeReferenceNode(node.parameters[0].type)
            && ts.isIdentifier(node.parameters[0].type.typeName)
            && node.parameters[0].type.typeName.text === "K"
        ) {
            return ""
        }

        return null
    },

    inject(node, context, render) {
        if (context.type !== "MEMBER") return []

        let name

        if (ts.isClassDeclaration(node) && node.name) {
            name = node.name
        } else if (ts.isInterfaceDeclaration(node)) {
            name = node.name
        }

        if (!name) return []

        const typeScriptService = context.lookupService(karakum.typeScriptServiceKey.get())
        const typeChecker = typeScriptService?.program.getTypeChecker()

        const symbol = typeChecker?.getSymbolAtLocation(name)
        if (!symbol) return []

        const events = this.events.get(symbol)
        if (!events) return []

        const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"
        const [, moduleOpenEvents] = Array.from(Object.entries(openEvents))
            .find(([fileName]) => sourceFileName.endsWith(fileName))
        ?? []
        const [, moduleOverriddenEvents] = Array.from(Object.entries(overriddenEvents))
            .find(([fileName]) => sourceFileName.endsWith(fileName))
        ?? []
        const [, moduleIgnoredEvents] = Array.from(Object.entries(ignoredEvents))
            .find(([fileName]) => sourceFileName.endsWith(fileName))
        ?? []

        return Array.from(events.entries())
            .map(([eventName, parameters]) => {
                const isOpen = moduleOpenEvents?.[name.text]?.includes(eventName) ?? false
                const isOverridden = moduleOverriddenEvents?.[name.text]?.includes(eventName) ?? false
                const isIgnored = moduleIgnoredEvents?.[name.text]?.includes(eventName) ?? false

                if (isIgnored) return ""

                let modifier = ""

                if (isOpen) {
                    modifier = "open "
                }

                if (isOverridden) {
                    modifier = "override "
                }

                const payload = parameters
                    .map(parameter => (
                        parameter.type
                            ? render(parameter.type)
                            : "Any?"
                    ))
                    .join(", ")

                const key = karakum.identifier(eventName).replaceAll("`", "")

                const tuple = parameters.length > 0
                    ? `js.array.JsTuple${parameters.length}<${payload}>`
                    : "js.array.JsTuple"

                return (
                    `
@web.events.JsEvent("${eventName}")
${modifier}val ${key}Event: node.events.EventInstance<${tuple}>
                    `.trim()
                );
            })
    },

    generate() {
        return []
    },
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/injections/EventMapInjection.kt
package node.karakum.injections

import io.github.sgrishchenko.karakum.extension.Context
import io.github.sgrishchenko.karakum.extension.GeneratedFile
import io.github.sgrishchenko.karakum.extension.Injection
import io.github.sgrishchenko.karakum.extension.InjectionContext
import io.github.sgrishchenko.karakum.extension.InjectionType
import io.github.sgrishchenko.karakum.extension.Render
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import io.github.sgrishchenko.karakum.util.camelize
import io.github.sgrishchenko.karakum.util.getParentOrNull
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import js.array.ReadonlyArray
import arrow.core.raise.impure
import arrow.core.raise.nullable
import typescript.DeclarationStatement
import typescript.Node
import typescript.ParameterDeclaration
import typescript.Symbol
import typescript.asArray
import typescript.isClassDeclaration
import typescript.isFunctionTypeNode
import typescript.isIdentifier
import typescript.isInterfaceDeclaration
import typescript.isIntersectionTypeNode
import typescript.isMethodDeclaration
import typescript.isMethodSignature
import typescript.isParameter
import typescript.isPropertySignature
import typescript.isTypeAliasDeclaration
import typescript.isTypeLiteralNode
import typescript.isTypeOperatorNode
import typescript.isTypeReferenceNode

private val eventHandlerMethods = setOf(
    "on",
    "off",
    "once",
    "emit",
    "addListener",
    "removeListener",
    "prependListener",
    "prependOnceListener",
    "listeners",
)

private val openEvents = mapOf<String, Map<String, Set<String>>>()

private val overriddenEvents = mapOf(
    "fs.d.ts" to mapOf(
        "ReadStream" to setOf(
            "close",
            "data",
            "end",
            "error",
            "pause",
            "readable",
            "resume",
        ),
        "WriteStream" to setOf(
            "close",
            "drain",
            "error",
            "finish",
            "pipe",
            "unpipe",
        ),
    ),
)

private val ignoredEvents = mapOf<String, Map<String, Set<String>>>()

private fun extractEventContainer(node: Node): DeclarationStatement? = nullable {
    ensure(isIdentifier(node))
    ensure(node.text == "K")

    val typeReference = ensureNotNull(node.getParentOrNull())
    ensure(isTypeReferenceNode(typeReference))

    val parameter = ensureNotNull(typeReference.getParentOrNull())
    ensure(isParameter(parameter))

    val parameterName = parameter.name
    ensure(isIdentifier(parameterName))
    ensure(parameterName.text == "event")

    nullable {
        val method = ensureNotNull(parameter.getParentOrNull())
        ensure(isMethodDeclaration(method))

        val methodName = ensureNotNull(method.name)
        ensure(isIdentifier(methodName))
        ensure(methodName.text in eventHandlerMethods)

        val classNode = ensureNotNull(method.getParentOrNull())
        ensure(isClassDeclaration(classNode))

        classNode
    } ?: nullable {
        val method = ensureNotNull(parameter.getParentOrNull())
        ensure(isMethodSignature(method))

        val methodName = ensureNotNull(method.name)
        ensure(isIdentifier(methodName))
        ensure(methodName.text in eventHandlerMethods)

        val interfaceNode = ensureNotNull(method.getParentOrNull())
        ensure(isInterfaceDeclaration(interfaceNode))

        interfaceNode
    }
}

private fun extractEventPayloads(node: Node, context: Context) = nullable {
    ensure(isIdentifier(node))
    ensure(node.text == "K")

    val typeReference = ensureNotNull(node.getParentOrNull())
    ensure(isTypeReferenceNode(typeReference))

    val parameter = ensureNotNull(typeReference.getParentOrNull())
    ensure(isParameter(parameter))

    val parameterName = parameter.name
    ensure(isIdentifier(parameterName))
    ensure(parameterName.text == "event")

    val method = ensureNotNull(
        nullable {
            val method = ensureNotNull(parameter.getParentOrNull())
            ensure(isMethodDeclaration(method))

            val methodName = ensureNotNull(method.name)
            ensure(isIdentifier(methodName))
            ensure(methodName.text == "addListener")

            val classNode = ensureNotNull(method.getParentOrNull())
            ensure(isClassDeclaration(classNode))

            method
        } ?: nullable {
            val method = ensureNotNull(parameter.getParentOrNull())
            ensure(isMethodSignature(method))

            val methodName = ensureNotNull(method.name)
            ensure(isIdentifier(methodName))
            ensure(methodName.text == "addListener")

            val interfaceNode = ensureNotNull(method.getParentOrNull())
            ensure(isInterfaceDeclaration(interfaceNode))

            method
        }
    )

    val typeParameter = method.typeParameters?.asArray()
        ?.find { it.name.text == "K" }
        .let { ensureNotNull(it) }

    val typeParameterConstraint = ensureNotNull(typeParameter.constraint)
    ensure(isTypeOperatorNode(typeParameterConstraint))

    val typeReferenceConstraintType = typeParameterConstraint.type
    ensure(isTypeReferenceNode(typeReferenceConstraintType))

    val typeScriptService = ensureNotNull(context.lookupService<TypeScriptService>(typeScriptServiceKey))
    val typeChecker = typeScriptService.program.getTypeChecker()

    val symbol = ensureNotNull(typeChecker.getSymbolAtLocation(typeReferenceConstraintType.typeName))

    val declaration = ensureNotNull(symbol.declarations?.firstOrNull())
    ensure(isTypeAliasDeclaration(declaration))

    val declarationType = declaration.type
    ensure(isIntersectionTypeNode(declarationType))

    val eventMap = declarationType.types.asArray()
        .find { isTypeLiteralNode(it) }
        .let { ensureNotNull(it) }
    ensure(isTypeLiteralNode(eventMap))

    eventMap.members.asArray()
        .mapNotNull { member ->
            nullable {
                ensure(isPropertySignature(member))

                val propertyName = member.name
                ensure(isIdentifier(propertyName))

                val propertyType = ensureNotNull(member.type)
                ensure(isFunctionTypeNode(propertyType))

                propertyName.text to propertyType.parameters.asArray()
            }
        }
        .toMap()
}

class EventMapInjection : Injection {
    private val events = mutableMapOf<Symbol, Map<String, ReadonlyArray<ParameterDeclaration>>>()

    override fun setup(context: Context) = Unit

    override fun traverse(node: Node, context: Context) = impure {
        ensureNotNull(isIdentifier(node))

        val eventContainer = ensureNotNull(extractEventContainer(node))
        val eventPayloads = ensureNotNull(extractEventPayloads(node, context))

        val name = ensureNotNull(eventContainer.name)

        val typeScriptService = ensureNotNull(context.lookupService<TypeScriptService>(typeScriptServiceKey))
        val typeChecker = typeScriptService.program.getTypeChecker()

        val symbol = ensureNotNull(typeChecker.getSymbolAtLocation(name))

        events[symbol] = eventPayloads
    }

    override fun render(node: Node, context: Context, next: Render<Node>) = nullable {
        val method = ensureNotNull(
            nullable {
                ensure(isMethodDeclaration(node))
                node
            } ?: nullable {
                ensure(isMethodSignature(node))
                node
            }
        )

        val name = ensureNotNull(method.name)
        ensure(isIdentifier(name))
        ensure(name.text in eventHandlerMethods)

        val firstParameter = ensureNotNull(method.parameters.asArray().firstOrNull())

        val firstParameterName = firstParameter.name
        ensure(isIdentifier(firstParameterName))
        ensure(firstParameterName.text == "event")

        val firstParameterType = ensureNotNull(firstParameter.type)
        ensure(isTypeReferenceNode(firstParameterType))

        val firstParameterTypeName = firstParameterType.typeName
        ensure(isIdentifier(firstParameterTypeName))
        ensure(firstParameterTypeName.text == "K")

        ""
    }

    override fun inject(node: Node, context: InjectionContext, render: Render<Node>) = nullable {
        ensure(context.type == InjectionType.MEMBER)

        val name = ensureNotNull(
            nullable {
                ensure(isClassDeclaration(node))
                ensureNotNull(node.name)
            } ?: nullable {
                ensure(isInterfaceDeclaration(node))
                node.name
            }
        )

        val typeScriptService = ensureNotNull(context.lookupService<TypeScriptService>(typeScriptServiceKey))
        val typeChecker = typeScriptService.program.getTypeChecker()

        val symbol = ensureNotNull(typeChecker.getSymbolAtLocation(name))

        val symbolEvents = ensureNotNull(events[symbol])

        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName

        val moduleOpenEvents = openEvents.entries
            .firstOrNull { (fileName) -> sourceFileName.endsWith(fileName) }
            ?.value

        val moduleOverriddenEvents = overriddenEvents.entries
            .firstOrNull { (fileName) -> sourceFileName.endsWith(fileName) }
            ?.value

        val moduleIgnoredEvents = ignoredEvents.entries
            .firstOrNull { (fileName) -> sourceFileName.endsWith(fileName) }
            ?.value

        symbolEvents.entries
            .map { (eventName, parameters) ->
                val isOpen = moduleOpenEvents?.get(name.text)?.contains(eventName) ?: false
                val isOverridden = moduleOverriddenEvents?.get(name.text)?.contains(eventName) ?: false
                val isIgnored = moduleIgnoredEvents?.get(name.text)?.contains(eventName) ?: false

                if (isIgnored) {
                    ""
                } else {
                    val modifier = when {
                        isOpen -> "open "
                        isOverridden -> "override "
                        else -> ""
                    }

                    val payload = parameters.joinToString(", ") { parameter ->
                        parameter.type?.let { render(it) } ?: "Any?"
                    }

                    val key = camelize(
                        eventName.replace("\\W".toRegex(), "-")
                    )

                    val tuple = if (parameters.isNotEmpty()) {
                        "js.array.Tuple${parameters.size}<${payload}>"
                    } else {
                        "js.array.Tuple"
                    }

                    """
                        @web.events.JsEvent("$eventName")
                        ${modifier}val ${key}Event: node.events.EventInstance<${tuple}>
                    """.trimIndent()
                }
            }
            .toTypedArray()
    }

    override fun generate(context: Context, render: Render<Node>) = emptyArray<GeneratedFile>()
}
